### PR TITLE
Security warning

### PR DESCRIPTION
**Pull request by PromyLOPh**

### CODE REVIEW

File: bawwab/client/userStorage.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+			console.error("Failed to create userStorage-file as a result of failing to find or create userStorage-directory!");
+			return false;
+		});
+}
+
+/**
+ * Provides basic means to read/write a key-value store (JSON-format) for a given user to the server.
+ * 
+ * Please remark that operations are not by themselves ACIDic and every read/write operation does their job
+ * with regards to the *whole* storage.
```
Comment (PromyLOPh) [inline bawwab/client/userStorage.js:100]: We could add a mashru3 helper function or locking inside of bawwab, if we wanted that.

File: bawwab/client/view/account.js [hunk 2]
```diff
@@ -42,21 +61,55 @@ export default {
...
+			const notAgainCB = document.getElementById('not-again');
+
+			autoCopyCB.addEventListener('change', () => {
+				this.settings.setAutocopySharedReadOnly(autoCopyCB.checked);
+			});
+
+			notAgainCB.addEventListener('change', () => {
+				this.settings.setDontShowSharedReadOnlyPopUp(notAgainCB.checked);
+			});
+		},
```
Comment (PromyLOPh) [inline bawwab/client/view/account.js:112]: Isn’t it possible to use Vue models here?

File: bawwab/client/view/workspaceExport.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ import { store } from '../app.js';
 export default {
 	name: 'WorkspaceExportView',
 	props: ['wsid'],
-	template: `<modal :title="t('headline')" :closeName="t('back')" :closeLink="{name: 'workspace', params: {wsid: workspace.metadata._id}}" :scaling="true" icon="file-export">
+	template: `<modal :title="t('headline')" :closeName="t('back')" :closeLink="{name: 'workspace', params: {wsid: workspace?.metadata._id}}" :scaling="true" icon="file-export">
```
Comment (PromyLOPh) [inline bawwab/client/view/workspaceExport.js:7]: Aw, this unrelated change should probably go into its own commit.

File: bawwab/client/view/workspaceSecurityPrompt.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
+import { translations, i18nMixin } from '../i18n.js';
+import { store } from '../app.js';
+import { Settings } from '../settings';
+
+export default {
+	props: ['wsid'],
+	template: `<modal icon="users" :title="t('title')" :closeLink="{name: 'workspace', params: {wsid: workspace?.metadata._id}}" :scaling="true">
+	<div class="share-prompt-info">
+		<i class="fas fa-question-circle inline-icon"></i>
+		<a class="info-link" href="https://www.psychnotebook.org/faq">{{ t('explainLink') }}</a>
```
Comment (PromyLOPh) [inline bawwab/client/view/workspaceSecurityPrompt.js:10]: Please use a router link.

File: bawwab/client/view/workspaceSecurityPrompt.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		isOwnWorkspace: function () { return this.workspace.owner().includes(this.username) },
+	},
+	methods: {
+		close: function() {
+			this.$router.push({name: 'workspace', params: {wsid: this.workspace?.metadata._id}});
+		},
+		loadSettings: async function() {
+			if (this.settingsAvailable) {
+				document.getElementById('autocopy').checked = await this.settings.getAutocopySharedReadOnly();
+				document.getElementById('not-again').checked = await this.settings.getDontShowSharedReadOnlyPopUp();
```
Comment (PromyLOPh) [inline bawwab/client/view/workspaceSecurityPrompt.js:108]: Same here, Vue models might make this easier, because you don’t have to touch the DOM. If you must, please use `this.$el` instead of `document`.

File: bawwab/client/view/workspaceShow.js [hunk 6]
```diff
@@ -127,12 +129,56 @@ export default {
...
 	mixins: [i18nMixin],
+	mounted: async function () {
+		if (!this.workspace || this.isOwnWorkspace) {
+			return;
+		}
+
+		this.settingsAvailable = await Settings.available();
+		let firstVisit = await this.firstVisitOnWorkspace();
+		let autocopy = await this.autocopy();
+		let dontShowPopUp = await this.dontShowPopUp();
```
Comment (PromyLOPh) [inline bawwab/client/view/workspaceShow.js:140]: These are constant, make them `const`.

File: bawwab/client/view/workspaceShow.js [hunk 7]
```diff
@@ -127,12 +129,56 @@ export default {
...
+		for projects that are not owned by the user */
+		if (!this.settingsAvailable && !this.isOwnWorkspace && firstVisit) {
+			await this.$router.push({ name: 'workspaceSecurityPrompt', params: { wsid: this.workspace.metadata._id } });
+			return;
+		}
+
+		if (
+			!this.settingsAvailable 
+			|| !firstVisit
+		) {
```
Comment (PromyLOPh) [inline bawwab/client/view/workspaceShow.js:152]: Coding style: Can be one line.

File: bawwab/client/view/workspaceShow.js [hunk 8]
```diff
@@ -148,14 +194,29 @@ export default {
 		canEditMeta: function () { return this.permissions.canWrite (); },
 	},
 	methods: {
-        save: async function () {
+		firstVisitOnWorkspace: async function() {
+			if (!this.settingsAvailable) {
+				return true;
+			}
+			return !await this.settings.getItem(`notYourFirstTimeOn${this.workspace.metadata._id}`);
```
Comment (PromyLOPh) [inline bawwab/client/view/workspaceShow.js:201]: You can turn the item name into a computed property, so you don’t need to spell it out twice (see below).

File: bawwab/client/settings.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+	async getItem(key) {
+		if (!await this.init()) {
+			return new Error("Settings-object not correctly initilized!");
+		}
+
+		let store = await this.userStorage.read();
+		console.log("User Storage: ", store);
+		return store[key];
+	}
```
Comment (PromyLOPh) [inline bawwab/client/settings.js:90]: I would make them synchronous by caching the settings locally on application load. Sure, this is even worse in terms of consistency, but you don’t need an extra HTTP request for every `getItem`, which will really kill responsiveness when navigating (since every project page uses it).

File: bawwab/client/settings.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
+import { UserStorage } from "./userStorage";
+
+/**
+ * Provides a means to set and get user-specific settings in form of key-value-pairs.
+ * 
+ * The current implementation is basically just a convenience-wrapper
+ * around UserStorage.
+ */
```
Comment (PromyLOPh) [inline bawwab/client/settings.js:8]: Settings+UserStorage is totally over-engineered. All you really need is a class mimicking a Mapping, which you can `async sync()` to disk.

File: bawwab/client/userStorage.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+ * 
+ * @param {string} username 
+ * @returns a promise resolving to true if the userStorage-directory has been succesfully created 
+ *   or was already present, resolves to false otherwise
+ */
+async function createDirectoryIfNotExistent(username) {
+	return fetch(`/api/filesystem/storage/home/${username}/.config/bawwab/userStorage/`, {
+		method: "GET",
+	})
+	.then(response => {
```
[the remark this answers is not in the document]
Comment (philepsybo) [reply bawwab/client/userStorage.js:11]: I think this was done to make those non-blocking but admittedly this is probably somewhat pointless as the user-code doesn't make too much use of this property.

File: bawwab/client/userStorage.js (hunk 1 above)
Comment (philepsybo) [reply bawwab/client/userStorage.js:100]: Kinda collides with the caching proposition for the settings that further corrupts consistency though I would argue in favor of caching instead of consistency as I don't see the importance of ACID with our limited use-cases so far.

File: bawwab/client/view/account.js (hunk 2 above)
Comment (philepsybo) [reply bawwab/client/view/account.js:112]: Maybe I don't understand your input correctly or I am missing the simple solution, but as I understand the docs, v-model doesn't bind naturally to computed/method. So this might need a workaround if we wanted to use v-model; see for example: https://riptutorial.com/vue-js/example/28978/using-computed-setters-for-v-model 

Though in this case I would argue that we don't get a real understandability-benefit this way. Another maybe a bit more vue-istic approach might be to make the relevant settings-values part of the data, v-model them to the checkboxes and setup @change-handlers to sync the changes back to the server on change. Once again I am not sure though if this would provide more clarity. Especially so if we decided to circumvent the Settings-API to access the internal map of the settings directly instead of using its methods (which might be necessary if we don't want to write get/set for both properties to make use of the properties, hence producing even more clutter).

File: bawwab/client/view/workspaceShow.js (hunk 6 above)
Comment (philepsybo) [reply bawwab/client/view/workspaceShow.js:140]: Addressed in 3be0fa3

File: bawwab/client/userStorage.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (philepsybo) [reply bawwab/client/userStorage.js:11]: Addressed in 77c297b

Not entirely happy though I must admit. After actually making the changes I feel like the code contains way too many `await`s ,i.e. blocking code. We should be careful not to make this an unnecessary pattern even if in this case the drawbacks probably turn out to be negligible.

File: bawwab/client/userStorage.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+ * 
+ * @param {string} username the unix-username of the person for whom a storage file should be created
+ * @returns a promise resolving to true if the storage file was successfulyy created,
+ *   resolves to false otherwise
+ */
+async function createStorageFile(username) {
+	return createDirectoryIfNotExistent(username)
+		.then(success => {
+			if (success) {
+				return fetch(`/api/filesystem/storage/home/${username}/.config/bawwab/userStorage/storage.json`, {
```
[the remark this answers is not in the document]
Comment (philepsybo) [reply bawwab/client/userStorage.js:78]: Addressed in 77c297b

File: bawwab/client/view/workspaceShow.js (hunk 7 above)
Comment (philepsybo) [reply bawwab/client/view/workspaceShow.js:152]: Addressed in 77c297b

File: bawwab/client/view/workspaceShow.js (hunk 8 above)
Comment (philepsybo) [reply bawwab/client/view/workspaceShow.js:201]: Addressed in 70cb2e7

File: bawwab/client/settings.js (hunk 9 above)
Comment (philepsybo) [reply bawwab/client/settings.js:90]: Addressed in 77c297b

File: bawwab/client/settings.js (hunk 10 above)
Comment (philepsybo) [reply bawwab/client/settings.js:8]: Addressed in 77c297b
